File: SDAT.pl
#!/usr/bin/env perl
# vim: ts=4 ai
# SDAT - Scanned document archival tool
#
#Copyright Ziva-Vatra, Belgrade
#(www.ziva-vatra.com, mail: info@ziva_vatra.com)
#
# Project REPO: https://github.com/ZivaVatra/SDAT
#
#Licensed under the GNU GPL. Do not remove any information from this header
#(or the header itself). If you have modified this code, feel free to add your
#details below this (and by all means, mail me, I like to see what other people
#have done)
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License (version 2)
#as published by the Free Software Foundation.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
#
# This is a script for archival of Documents/Bills/Invoices/etc...
# It scans the page, runs OCR on the text, and saves the text to the comment field in the metadata
# This allows indexing engines (e.g. Desktop search) to know what text the document contains, allowing
# for easier searching, while keeping the original text+format as an image scan.
# It saves a PNG file into the $FINALDST folder

# Requirements:
#	tesseract (OCR)
#	sane-tools (SCANNING)
#	Exiv2 image metadata library (for adding text to comment field)
#	imageMagick tools (FORMAT CONVERSION)
#

use strict;
use lib "./";
use SDAT::core;

use POSIX "sys_wait_h";
use File::Path "make_path";
use File::Basename "fileparse";

# Global defaults
our $SCAN_DPI=0;
# Extra options for scanimage, for specific scanners
our $DEVICE;
our @EXTRAOPTS;
our $HAS_ADF;
our $OUTFORMAT = "null";
# Extra options for tesseract
our @TESSOPTS=("-l", "eng");
our $ENABLE_DUPLEX=1;
# We define this here so that it is available in the config file.
# We don't assign a value however until we have the "scanCore" class
# instantiated, after which we know what the temp path will be
our $TEMPDIR; 
# options
our $NO_OCR = 0;
if (defined($ENV{'SDAT_NO_OCR'})) {
	$NO_OCR = $ENV{'SDAT_NO_OCR'};
}
our $OCR_ENABLED = not $NO_OCR; # backwards compatibility, NO_OCR is deprecated and will be removed in future

sub usage {
	die("Usage: $0 \$configuration_file \$scanner_file \$target_folder \$target_scan_filename\n");
}
my $CONF_FILE=shift or usage(); # Configuration
my $SPROFILE=shift or usage(); # Scanner profile
our $FINALDST=shift or usage(); #destination (global for callback)
our $NAME=shift or usage(); #filename (global for callback)

# Read in config file
die("Not a valid config file ($!)\n") unless ( -f $CONF_FILE);
require $CONF_FILE;
die "Couldn't interpret the configuration file ($CONF_FILE) that was given.\nError details follow: $@\n" if $@;

die("Not a valid scanner profile ($!)\n") unless (-f $SPROFILE);
require $SPROFILE;
die "Couldn't interpret the scanner profile file ($SPROFILE) that was given.\nError details follow: $@\n" if $@;

# If DPI is still "0", we have invalid config, so cannot continue
die("Invalid configuration file detected, cannot continue.\n") if $SCAN_DPI == 0;
die("Invalid scanner profile detected, cannot continue.\n") unless defined($DEVICE);



my $RANDSTR=`head -c 20 /dev/urandom  | md5sum | cut -d' ' -f 1`;
$RANDSTR =~ s/\n//g;
$NAME =~ s/\n//g;

my $scanCore = SDAT::core->new({
	"resolution" => $SCAN_DPI,
	"outDIR" => $FINALDST,
	"filePattern" => $NAME,
	"scanOpts" => \@EXTRAOPTS,
	"device" => $DEVICE,
	"tessOpts" => \@TESSOPTS,
	"OCR" => $OCR_ENABLED,
	"hasADF" => $HAS_ADF,
	"duplex" => $ENABLE_DUPLEX,
	"outFormat" => $OUTFORMAT,
	});

# Assigned so that it is available to callback_last;
$TEMPDIR = $scanCore->{tempDIR};

# 1. Scan the images to a temporary folder
# As this can take a while, we fork
my $pid = fork();
if ($pid == 0) {
	$scanCore->scan();
	exit;
}

#While the above is scanning, we sit and wait for files to be created,
# Then process them as they arrive
my $counter = 3;
while(1) {
	sleep(1);

	# Loop through images, for each one do the OCR, and move to dest
	my @outfiles = glob("$scanCore->{tmpDIR}/$scanCore->{filePattern}*.tiff");
	# As long as the scanning pid is not dead, reset
	# counter
	if (waitpid($pid, WNOHANG) != -1) {
		$counter = 3;
	}
	if(@outfiles) {
		Forks::Super::pmap { $scanCore->OCR($_) } {timeout => 120}, @outfiles;
	}
	
	print("Waiting for processing pid\n");

	if ($counter-- <= 0) {
		print(" Finished!\n");
		# We have reached end of countdown with no files
		# and dead scanning PID, quit loop
		last;
	}
	Forks::Super::waitall(); # We wait for any children to finish their processing
}
# Finally, we check to see if callback_last function is defined. If it is, we execute
if (defined(&callback_last)) {
	callback_last();
}

# When all processing is complete, we write out our PNGs or PDFs
$scanCore->writeFormatBatch() or die("Could not write output!");

sub process {
	my $inFile = shift;
}
# When all is done, remove tmp folder
#exe("rm -rv $TPATH/scan.$RANDSTR");

File: SDAT.py
#!/usr/bin/env python2
# vim: ts=4 ai
# SDAT - Scanned document archival tool
#
# Copyright 2017 Ziva-Vatra, Belgrade
# (www.ziva-vatra.com, mail: info@ziva_vatra.com)
#
# Project URL: http://www.ziva-vatra.com/index.php?aid=71&id=U29mdHdhcmU=
# Project REPO: https://github.com/ZivaVatra/SDAT
#
# Licensed under the GNU GPL. Do not remove any information from this header
# (or the header itself). If you have modified this code, feel free to add your
# details below this (and by all means, mail me, I like to see what other
# people have done)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License (version 2)
# as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
# USA.
#
#
# This is a script for archival of Documents/Bills/Invoices/etc...
# It scans the page, runs OCR on the text, and saves the text to the comment
# field in the metadata
# This allows indexing engines (e.g. Desktop search) to know what text the
# document contains, allowing
# for easier searching, while keeping the original text+format as an image
# scan.
# It saves a PNG file into the $FINALDST folder

# Requirements:
#  tesseract (OCR)
#  sane-tools (SCANNING)
#  Exiv2 image metadata library (for adding text to comment field)
#  imageMagick tools (FORMAT CONVERSION)
#

import os
import subprocess
import sys
import uuid


def get_device():
	if os.path.exists("./devicename") is True:
		with open("./devicename", 'r') as fd:
			return fd.read()

	data = subprocess.check_output(['scanimage', '-L'])
	data = data.split(' ')[1]
	data = data.replace('`', '')
	data = data.replace('\'', '')
	with open("./devicename", "w") as fd:
		fd.write(data)
	return data


def exe(cmd):
	return subprocess.check_call(cmd)


def bgexe(cmd):
	return os.spawnl(os.P_NOWAIT, cmd)


def scanit(mode, resolution, output, fail=False):

	print "Waiting for scanning to finish"
	try:
		data = subprocess.check_output([
			"scanimage",
			"-v",
			"-p",
			"--format=tiff",
			"--mode", mode,
			"-d", get_device(),
			"--resolution",
			str(resolution)
		])
	except subprocess.CalledProcessError as e:
		if fail is False:
			print "Got Error %s. Attempting to re-detect scanner" % e
			os.unlink("./devicename")
			get_device()
			return scanit(mode, resolution, output, True)
		else:
			print "Still unable to scan. Aborting execution."
			raise(e)

	with open(output, 'w') as fd:
		fd.write(data)


def s600dpi_col(output):
	scanit("color", 600, output)

def s1200dpi_col(output):
	scanit("color", 1200, output)

def s600dpi_gr(output):
	scanit("gray", 600, output)


def s1200dpi_gr(output):
	scanit("gray", 1200, output)


def ocrit(input_image, output_text_file, lang="eng"):
	return exe([
		"tesseract",
		input_image,
		output_text_file.replace('.txt', ''),
		"--tessdata-dir",
		"/usr/share/tesseract-ocr/",
		"-l",
		lang
	])


def topng(input, output):
	exe([
		"convert",
		"-compress",
		"Zip",
		input,
		output
	])


def addComment(input_file, output_image):
	with open(input_file, 'r') as fd:
		input_text = fd.read()

	exe([
		"exiv2",
		"-M set Exif.Photo.UserComment charset=Ascii %s" % input_text,
		output_image
	])


def waituntildone(pid):
	return os.waitpid(pid, 0)

if __name__ == "__main__":
	TPATH = "/tmp/scanning/"
	NAME = sys.argv[1]  # filename, first argument given to script
	device = get_device()
	print "Using Device: ", device

	RANDSTR = str(uuid.uuid4())
	FINALDST = sys.argv[2]
	if not os.path.exists(FINALDST):
		os.mkdir(FINALDST)

	print("Scanning filename: %s (Final destination: %s), CTRL-C to cancel" % (
		NAME, FINALDST
	))

	if not os.path.exists(TPATH):
		os.mkdir(TPATH)

	# 1. Scan the image (gray for OCR)
	grayfile = os.path.join(TPATH, "scan_gr", RANDSTR + ".tif")

	if not os.path.exists(os.path.join(TPATH, "scan_gr")):
		os.mkdir(os.path.join(os.path.join(TPATH, "scan_gr")))

	if not os.path.exists(os.path.join(TPATH, "scan_col")):
		os.mkdir(os.path.join(TPATH, "scan_col"))

	s1200dpi_gr(grayfile)

	# 2. Scan the colour image we will store
	colfile = os.path.join(TPATH, "scan_col", RANDSTR + ".tif")
	s1200dpi_col(colfile)

	textfile = os.path.join(TPATH, "scan_txt" + RANDSTR + ".txt")
	ocrpid = ocrit(grayfile, textfile)

	# Right, the below needs no user input, so we can just fork and exit the
	# program.
	# this way we can go to the next scan while this does work in the background

	pid = os.fork()
	if (pid == 0):
		# ok, all done! Now convert to png and add text
		topng(colfile, os.path.join(TPATH, NAME + ".png"))
		outfile = os.path.join(TPATH, NAME + ".png")
		addComment(textfile, outfile)
		exe(["mv", "-v", outfile, FINALDST])
		subprocess.check_call("rm -v " + TPATH + "/*%s*" % RANDSTR, shell=True)
	else:
		bgexe([
			"display",
			"-sample",
			750,
			FINALDST,
			NAME + ".png"
		])
		exit(0)
